yt_dlp/extractor/graspop.py

```diff
@@ -0,0 +1,42 @@
+from .common import InfoExtractor
+from ..utils import (
+    traverse_obj,
+)
+
+
+class GraspopIE(InfoExtractor):
+    _VALID_URL = r'https?://vod\.graspop\.be/(?P<lang>fr|nl)/(?P<id>[0-9]+)/(?P<title>.*)/'
+    _TESTS = [{
+        'url': 'https://vod.graspop.be/fr/101556/thy-art-is-murder-concert/',
+        'info_dict': {
+            'id': '101556',
+            'ext': 'mp4',
+            'title': 'Thy Art Is Murder',
+            'description': 'Thy Art Is Murder @ Graspop',
+            'thumbnail': r're:https://cdn-mds\.pickx\.be/festivals/v3/global/original/.*\.jpg',
+            'formats': 'count:4',
+        },
+        'params': {
+            'nocheckcertificate': True,
+        },
```

Comment:
First, thank you for your quick and thorough review (the inline return with advanced `traverse_obj` is nice!)

---

Alas, the `assetUrl` redirects to a server that doesn't serve the full certificate path, breaking the extractor at least on my linux machine:

```
$ yt-dlp https://vod.graspop.be/fr/101556/thy-art-is-murder-concert/
[Graspop] Extracting URL: https://vod.graspop.be/fr/101556/thy-art-is-murder-concert/
[Graspop] 101556: Downloading JSON metadata
[Graspop] 101556: Downloading m3u8 information
ERROR: [Graspop] 101556: Failed to download m3u8 information: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1006) (caused by CertificateVerifyError('[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1006)')); please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U
```

I reproduce the issue with `curl`:

```
$ curl -s https://tv.proximus.be/MWC/videocenter/festivals/101556/stream | jq '.source.assetUri' -r | xargs curl -L
curl: (60) SSL certificate problem: unable to get local issuer certificate
More details here: https://curl.se/docs/sslcerts.html

curl failed to verify the legitimacy of the server and therefore could not
establish a secure connection to it. To learn more about this situation and
how to fix it, please visit the web page mentioned above.
```

(also reproduced on ubuntu 20.04 & 22.04)

https://www.ssllabs.com/ssltest/analyze.html?d=da-c3ee1e8b.id.cdi.streaming.proximustv.be&hideResults=on

Here they say the intermediate certificate in the chain is `Extra download`, I'm not familiar with what it is, but it doesn't seem supported in python & curl...

---

I' don't know why the tests seem to work now without `nocheckcertificate`: 
At some point in the development, I had to add `nocheckcertificate`.

The current version here pass tests, but adding `import pdb; pdb.set_trace()` before the `return` in `_real_extract` shows again the error:
```
test/test_download.py::TestDownload::test_Graspop 
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> PDB set_trace (IO-capturing turned off) >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
> /media/data/1000-2/dev/yt-dlp/yt_dlp/extractor/graspop.py(25)_real_extract()
-> 'id': video_id,
(Pdb) n
> /media/data/1000-2/dev/yt-dlp/yt_dlp/extractor/graspop.py(26)_real_extract()
-> 'formats': self._extract_m3u8_formats(metadata['source']['assetUri'], video_id, 'mp4'),
(Pdb) 
yt_dlp.utils.ExtractorError: Failed to download m3u8 information: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1006) (caused by CertificateVerifyError('[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1006)')); please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U
> /media/data/1000-2/dev/yt-dlp/yt_dlp/extractor/graspop.py(26)_real_extract()
-> 'formats': self._extract_m3u8_formats(metadata['source']['assetUri'], video_id, 'mp4'),
```


Other strange behavior: adding a broken check in `TESTS`, such as `'formats': 'count:50',` *does not* break the test, but it *is* breaking the test when adding `.replace('https://', 'http://')`.
Maybe the certificate exception during the tests bubbles-up badly and somehow skips the rest of the test *without error*?
-> confirmed, continuing during the pdb session when the exception is raised ends with test passed!
why?

---

To conclude:
- I don't really know why I get a certificate error when everything is fine in browsers
- I don't know why the tests pass when the certificate exception *is* raised
- Should I go with `.replace('https://', 'http://')` to avoid any other certificate issue, even if it hides from the end-user the security implications? It does require the end-user to google a bit to find `--no-check-certificate`, but at least they know what they do.

Reply:
I can repro

> I don't really know why I get a certificate error when everything is fine in browsers

Apparently, it's long-standing issue with python (and many curl builds): https://github.com/python/cpython/issues/62817

> I don't know why the tests pass when the certificate exception is raised

If you call the test script directly with `python test/test_download.py TestDownload.test_Graspop`, you'll see that it attempts 3 times and then outputs a warning that the test is being skipped. Truncated log:

```
Retrying: 2 failed tries

##########

[Graspop] Extracting URL: https://vod.graspop.be/fr/101556/thy-art-is-murder-concert/
[Graspop] 101556: Downloading JSON metadata
[Graspop] 101556: Downloading m3u8 information
ERROR: [Graspop] 101556: Failed to download m3u8 information: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1000) (caused by CertificateVerifyError('[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1000)')); please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U

yt_dlp.networking.exceptions.CertificateVerifyError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1000)

WARNING: test_Graspop failed due to network errors, skipping...
.
----------------------------------------------------------------------
Ran 1 test in 3.874s

OK

```

pytest, on the other hand, swallows the retry/warning output and just gives an "OK." This is not ideal. Will need to investigate what to do about that. It's nothing to be concerned about w/ this PR though.

> Should I go with .replace('https://', 'http://') to avoid any other certificate issue, even if it hides from the end-user the security implications? It does require the end-user to google a bit to find --no-check-certificate, but at least they know what they do.

Yeah, I think we should downgrade the m3u8 request, I've added a suggestion below